### ISSUE TITLE
Roslyn ignores inferred nullability of 'ref' arguments when inferring type arguments of an invocation

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#37909]
**Version Used**: 

```
Branch master (10 Aug 2019)
Latest commit c21548 by dotnet-automerge-bot:
Merge pull request #37896 from dotnet/merges/release/dev16.3-preview2-to-master

Merge release/dev16.3-preview2 to master
```

**Steps to Reproduce**:

```cs
#nullable enable

class C
{
  T M1<T>(ref T t) => t;
  T M2<T>(T t) => t;
  T M3<T>(in T t) => t;

  void Test1(string? s1)
  {
    if (s1 == null) return;
    
    s1.ToString(); // non nullable
    var t = M1(ref s1);
    t.ToString(); // nullable, CS8602
  }
    
  void Test2(string? s1)
  {
    if (s1 == null) return;
    
    s1.ToString(); // non nullable
    var t = M2(s1);
    t.ToString(); // non nullable
  }
    
  void Test3(string? s1)
  {
    if (s1 == null) return;
    
    s1.ToString(); // non nullable
    var t = M3(in s1);
    t.ToString(); // non nullable
  }
}
```

**Expected Behavior**:
No warnings.
All methods are inferred with type argument `string!`. All variables `t` are non-nullable.

**Actual Behavior**:
`warning CS8602: Dereference of a possibly null reference.` is reported for `t.ToString()` in `Test1` only.

@jcouv Could you please clarify if this is intended and if so what are the reasons for current behavior? It seems rather counterintuitive.

**Comment by jcouv** [comment]
From discussion with Neal, this behavior seems desirable. Consider the following scenario:
```csharp
string? s = "";
Copy(ref s, null);
// where void Copy<T>(ref T t, T t2)
```
For purpose of method type inference, if we count the `ref s` argument as a `string!` (because its current value is non-null), then we'll infer `string!` for `T` and get a warning on the `null` argument.

**Comment by jcouv** [comment]
To answer @jaredpar's question more generally, by-value and `in` arguments contribute their r-value to type inference, and `ref` and `out` arguments contribute their l-value. Aside from that, there are some tweaks for lambdas and `out` vars.

**Comment by jcouv** [comment]
Added tests to record current behavior. Closing issue.
